### ISSUE TITLE
Best practise for legacy classes

### ISSUE THREAD

**Issue by fritzmg** [contao/docs#100]
As discussed in slack, we should provide a best-practise documentation for the handling of legacy classes. e.g. that they should all reside in a single sub-namespace, like `Contao`, so that you can easily exclude them from service resource loading like so:

```yaml
services:
    App\:
        resource: ../src
        exclude: ../src{Contao,DependencyInjection,Resources,Entity}
```

Background: if you accidentally auto-register a class that extends from `\Contao\System`, you will run into a fatal error during container building.

**Comment by m-vo** [comment]
see https://github.com/symfony/symfony/issues/34053

**Comment by m-vo** [comment]
Services are a concept that works well with composition + interfaces. The legacy code requires you to inherit from all sorts of classes instead. Plus: constructor injection often isn't possible.

Considering this I think a **separate sub namespace** (that is excluded from service resource loading) for these classes should be the best practice no matter the outcome of the related Symfony issue. The same is true for tests and coverage. I would exclude these classes as well as they aren't testable for the same reason.

So you would end up with `App\Contao\Widget\MyWidget` instead of `App\Widget\MyWidget` for example.

**Comment by fritzmg** [comment]
I would like to get the approval of @contao/developers on this. So in the future, the recommended namespaces would be:

* `App\Contao\Model`
* `App\Contao\Widget`
* `App\Contao\ContentElement`
* `App\Contao\FrontendModule`

etc. 

This would also need to be reflected in the Skeleton Bundle by @leofeyer

**Comment by leofeyer** [comment]
No, there must not be a `Contao` namespace.

* `App\Model`
* `App\Widget`
* `App\ContentElement`
* `App\FrontendModule`

**Comment by Toflar** [comment]
It depends. I don't think I would like to give a recommendation here. People may still use their brains 😄
If I have an app where I only extend Contao then I would not add this subnamespace. It's useless. If I have an app that consists of content management and an API I would probably make a difference on top level and have additional `Contao` and `Api` namespaces.

**Comment by fritzmg** [comment]
> No, there must not be a `Contao` namespace.

Then the service resource-loading would need to be defined as

```yml
services:
    App\:
        resource: ../src
        exclude: ../src{Widget,Model,ContentElement,FrontendModule,DependencyInjection,Resources,Entity}
```

Plus any other subnamespace of legacy Contao classes.

**Comment by aschempp** [comment]
- `App\ContentElement`
- `App\FrontendModule` 

These should be controllers, so either `App\Controller` or `App\Controller\ContentElement` accordingly.

**Comment by fritzmg** [comment]
Of course, this is only in case you absolutely have to use the legacy way.

**Comment by aschempp** [comment]
Then I agree with @Toflar, there's no recommendation for the old way, because it's outdated 😂 
Yeah, people still need it, but there's only a good new way, there's no good old way. So do whatever you feel like …

**Comment by fritzmg** [comment]
So there is nothing that should be changed in the official documentation and people should deal with problems arising from loading legacy Contao classes as services on their own? Even if https://github.com/symfony/symfony/issues/34053 gets fixed in Symfony 3 & 4, I still think it would be good to show that it is best practise _not_ to load any Contao legacy classes as services.

**Comment by m-vo** [comment]
> Then I agree with @Toflar, there's no recommendation for the old way, because it's outdated joy

Not all things are outdated. How would you create a `Widget` for example?

**Comment by aschempp** [comment]
> I still think it would be good to show that it is best practise _not_ to load any Contao legacy classes as services.

I think thats a good idea as well. But how's that related to the namespace suggestion you were looking for? 

If this is about the _automatic_ service loading, then we should tell people to make sure they only load directories that contain service classes. That might include Doctrine entities, and it includes Contao 3 widgets, or any other class where there are factories for.

**Comment by m-vo** [comment]
The original issue (undefined constants) has been fixed in the Symfony 3.4 branch.

See https://github.com/symfony/symfony/pull/34560

**Comment by fritzmg** [comment]
This is now pretty much solved.

* The Contao Managed Edition has an automatic autoloading definintion for the `App\` namespace under `src/`.
* The Contao Managed Edition has automatic service registration for the `App\` namespace under `src/`.
* No errors should occur anymore even if you have classes extending from legacy classes within `src/`.
